### PR TITLE
feat(FEC-9678): add response filter implementation to hls adapter

### PR DESCRIPTION
**Pull request by yairans**
### Description of the Changes

Add `responseFilter` implementation:
- Override the `readystatechange` to apply the filter
- Consider filter error on `ERROR` event
- Do not apply retry when request/response filter throws an error
- Fix incorrect `pLoader.redirectExternalStreamsHandler` singture (used in tests)

Depends on https://github.com/kaltura/playkit-js/pull/449

Solves FEC-9678

### CheckLists

- [x] changes have been done against master branch, and PR does not conflict
- [x] new unit / functional tests have been added (whenever applicable)
- [x] test are passing in local environment
- [x] Travis tests are passing (or test results are not worse than on master branch :))
- [ ] Docs have been updated

### CODE REVIEW

File: src/hls-adapter.js [hunk 1]
```diff
@@ -120,6 +120,7 @@ export default class HlsAdapter extends BaseMediaSourceAdapter {
   _onMediaAttached: Function;
   _mediaAttachedPromise: Promise<*>;
   _requestFilterError: boolean = false;
+  _requestFilterError: boolean = false;
```
Comment (RoyBregman) [inline src/hls-adapter.js]: duplicate definition

Comment (RoyBregman) [inline src/hls-adapter.js]: Now i see it should have been _responseFilterError

File: src/hls-adapter.js [hunk 2]
```diff
@@ -972,12 +1074,14 @@ export default class HlsAdapter extends BaseMediaSourceAdapter {
...
-        : HlsJsErrorMap[errorName] || {category: 0, code: 0};
+      const {category, code}: ErrorDetailsType =
+        this._requestFilterError || this._responseFilterError
+          ? {category: Error.Category.NETWORK, code: this._requestFilterError ? Error.Code.REQUEST_FILTER_ERROR : Error.Code.RESPONSE_FILTER_ERROR}
+          : HlsJsErrorMap[errorName] || {category: 0, code: 0};
       HlsAdapter._logger.warn(new Error(Error.Severity.RECOVERABLE, category, code, errorDataObject));
     }
-    this._requestFilterError = false;
+    this._requestFilterError && (this._requestFilterError = false);
+    this._responseFilterError && (this._responseFilterError = false);
```
Comment (RoyBregman) [inline src/hls-adapter.js]: isn't this just same as:
this._requestFilterError = false
this._responseFilterError = false
Why do we need to check if it is true before?

File: src/jsonp-ploader.js [hunk 3]
```diff
@@ -10,11 +10,12 @@ import loader from './loader';
 export default class pLoader extends loader {
   /**
    * redirect external stream handler function
+   * @param {Object} data - The json object that returns from the server.
    * @param {string} uri - the original uri
    * @returns {string} uri - the redirected URI
    * @static
    */
-  static redirectExternalStreamsHandler: Function = uri => uri;
+  static redirectExternalStreamsHandler: Function = (data, uri) => uri;
```
Comment (RoyBregman) [inline src/jsonp-ploader.js]: i didn't find where this was changed

File: test/src/hls-adapter.spec.js [hunk 4]
```diff
@@ -732,6 +755,252 @@ describe('HlsAdapter Instance request filter', () => {
...
+                reject(new window.Error('error'));
+              });
+            }
+          }
+        }
+      })
+    );
+    hlsAdapterInstance.addEventListener(EventType.ERROR, event => {
+      try {
+        if (event.payload) {
```
Comment (RoyBregman) [inline test/src/hls-adapter.spec.js]: payload should always exist now. I think this is not relevant now

File: test/src/hls-adapter.spec.js [hunk 5]
```diff
@@ -732,6 +755,252 @@ describe('HlsAdapter Instance request filter', () => {
...
+            return new Promise(() => {
+              throw new window.Error('error');
+            });
+          }
+        }
+      })
+    );
+    hlsAdapterInstance.addEventListener(EventType.ERROR, event => {
+      try {
+        if (event.payload) {
```
Comment (RoyBregman) [inline test/src/hls-adapter.spec.js]: same here, no need to check if payload exists

File: src/jsonp-ploader.js (hunk 3 above)
Comment (yairans) [reply src/jsonp-ploader.js]: it wasn't changed. it was incorrect from begin. see [here](https://github.com/kaltura/playkit-js/blob/master/src/utils/jsonp.js#L51).
in real playback we overriding this function as u can see [here](https://github.com/kaltura/playkit-js-hls/blob/master/src/hls-adapter.js#L138). so it impacts the test only

File: src/hls-adapter.js [hunk 6]
```diff
@@ -943,11 +1030,19 @@ export default class HlsAdapter extends BaseMediaSourceAdapter {
...
+              code = Error.Code.RESPONSE_FILTER_ERROR;
+            } else {
+              code = Error.Code.HTTP_ERROR;
+            }
             if (
               [Hlsjs.ErrorDetails.MANIFEST_LOAD_ERROR, Hlsjs.ErrorDetails.MANIFEST_LOAD_TIMEOUT].includes(errorName) &&
               !this._triedReloadWithRedirect &&
-              !this._config.forceRedirectExternalStreams
+              !this._config.forceRedirectExternalStreams &&
+              !(this._requestFilterError || this._responseFilterError)
```
Comment (RoyBregman) [inline src/hls-adapter.js]: i think maybe it is easier to read it like this:
!this._requestFilterError && !this._responseFilterError, no?

**Pull request merged**